### PR TITLE
docs(plan): same-batch tool-fix ordering hazard detection (#199)

### PR DESCRIPTION
**Pull request by lhpaul**
## Summary

Implementation plan for issue #199 — detecting the same-batch tool-fix ordering hazard in Protocol 90 Step 3 and `workflow-batch-plan.sh`.

**Approach**: Two targeted changes:
1. `scripts/development-workflow/workflow-batch-plan.sh` — add `classify_tool_fix` helper that scans each candidate's spec/plan document for references to the canonical tool file set and emits `TOOL_FIX=yes|no|unknown` + `TOOL_FIX_FILES=...` per item.
2. `docs/ai/development-workflow/protocols/90-batch-orchestrate-work-protocol.md` Step 3 — add "Same-batch tool-fix ordering hazard" subsection covering: definition, detection sources (script + tracker), serialize-first rule, `unknown` handling, multiple-tool-fix ordering, already-waiting detection, human override path, and batch summary annotation.

**Estimated complexity**: S (< 1 day)

**Key risks**: Exact-path grep matching must not match superstrings; protocol glob matching for `docs/ai/development-workflow/protocols/*.md` requires a two-part approach (prefix + `.md` suffix pattern).

**Scope**: Only `workflow-batch-plan.sh` and Protocol 90 Step 3. Step 5.1 and all other protocols are out of scope.

**Links**:
- Plan: `docs/specs/developments/20260417154720_199-same-batch-tool-fix-ordering/2_199-same-batch-tool-fix-ordering_implementation-plan.md`
- Smoke test: `docs/testing/workflow/199-same-batch-tool-fix-ordering.smoke-test.md`

## Summary by CodeRabbit

* **Documentation**
  * Added an implementation plan describing detection and handling of same-batch "tool-fix" ordering hazards, including required workflow outputs and protocol updates for serialize-first ordering and human override annotations.
  * Added a smoke test runbook validating tool-fix detection outcomes (yes/no/unknown), exact-path matching behavior, protocol Step 3 content, and cleanup/troubleshooting steps.

### CODE REVIEW

File: docs/specs/developments/20260417154720_199-same-batch-tool-fix-ordering/2_199-same-batch-tool-fix-ordering_implementation-plan.md [hunk 1]
```diff
@@ -0,0 +1,247 @@
...
+  6. Emits `TOOL_FIX=yes` and `TOOL_FIX_FILES=<list>` when at least one canonical path is
+     matched, `TOOL_FIX=no` when none are matched, or `TOOL_FIX=unknown` when no spec/plan
+     document exists.
+
+- [ ] In `workflow-batch-plan.sh`, call `classify_tool_fix` for each development path inside
+  the main loop (after the existing `print_kv PARALLEL_SAFE` line) and emit the result using
+  `print_kv`:
+  - `print_kv TOOL_FIX "$tool_fix"` always.
+  - `print_kv TOOL_FIX_FILES "$tool_fix_files"` only when `TOOL_FIX=yes` (omit the line
+    entirely for `no` and `unknown` to keep output compact).
```
[the remark this answers is not in the document]
Comment (lhpaul) [reply docs/specs/developments/20260417154720_199-same-batch-tool-fix-ordering/2_199-same-batch-tool-fix-ordering_implementation-plan.md]: Addressed in commit `f405ec4` — plan now instructs placing `classify_tool_fix` before the `workflow-next-action.sh` guard, with an abbreviated output block emitted in the next-action-failed path so `TOOL_FIX=unknown` remains reachable for directories without spec/plan files (spec AC 13 now satisfiable).

File: docs/testing/workflow/199-same-batch-tool-fix-ordering.smoke-test.md [hunk 2]
```diff
@@ -0,0 +1,267 @@
...
+### Step 9: AC3 — Exact-path matching (no false positives from superstrings)
+
+Verify that a path like `scripts/development-workflow/pr-review-loop.sh.bak` does NOT trigger
+`TOOL_FIX=yes`:
+
+```bash
+mkdir -p .tmp/smoke-199/20200101000009_ac3-superstring
+cat > .tmp/smoke-199/20200101000009_ac3-superstring/1_ac3b_specs.md <<'EOF'
+# References pr-review-loop.sh.bak (a superstring, not the canonical path)
+EOF
```
[the remark this answers is not in the document]
Comment (lhpaul) [reply docs/testing/workflow/199-same-batch-tool-fix-ordering.smoke-test.md:178]: Addressed in commit `f405ec4` — fixture now contains the full canonical path with `.bak` suffix (`scripts/development-workflow/pr-review-loop.sh.bak`), so a naive substring match would incorrectly match, making the superstring test non-trivial.

File: docs/specs/developments/20260417154720_199-same-batch-tool-fix-ordering/2_199-same-batch-tool-fix-ordering_implementation-plan.md [hunk 3]
```diff
@@ -0,0 +1,254 @@
...
+     repo-relative path as written in the canonical list — not a substring of a longer path.
+     Implement using `grep -F` (fixed-string) against each canonical path individually so that
+     (for example) `pr-review-loop.sh` does not match `pr-review-loop.sh.bak` or
+     `docs/ai/.../protocols/90-batch-orchestrate-work-protocol.md` matching the glob
+     `docs/ai/development-workflow/protocols/*.md` is handled as an anchored prefix + suffix
+     check, not as a raw glob expansion.
+  4. For the glob `docs/ai/development-workflow/protocols/*.md`, match lines that contain a
+     path starting with `docs/ai/development-workflow/protocols/` and ending with `.md` (using a
+     grep pattern such as `docs/ai/development-workflow/protocols/[^/]*\.md`).
+  5. Collects all matched canonical paths into a comma-separated list.
```
[the remark this answers is not in the document]
Comment (lhpaul) [reply docs/specs/developments/20260417154720_199-same-batch-tool-fix-ordering/2_199-same-batch-tool-fix-ordering_implementation-plan.md]: Addressed in commit `0f58261` — plan text and code sample now require delimiter-aware matching. Both exact-path and protocols-glob matchers use `(^|[^[:alnum:]_./-])...($|[^[:alnum:]_./-])` boundaries, so superstrings like `pr-review-loop.sh.bak` are rejected.

File: docs/specs/developments/20260417154720_199-same-batch-tool-fix-ordering/2_199-same-batch-tool-fix-ordering_implementation-plan.md [hunk 4]
```diff
@@ -0,0 +1,254 @@
...
+  doc_file="$(find "$dev_path" -maxdepth 1 -name '*.md' | sort | head -1)"
+  if [ -z "$doc_file" ]; then
+    printf 'unknown\n'
+    return 0
+  fi
+
+  matched_paths=()
+
+  for path in "${CANONICAL_EXACT_PATHS[@]}"; do
+    if grep -qF "$path" "$doc_file"; then
```
[the remark this answers is not in the document]
Comment (lhpaul) [reply docs/specs/developments/20260417154720_199-same-batch-tool-fix-ordering/2_199-same-batch-tool-fix-ordering_implementation-plan.md]: Addressed in commit `0f58261` — code sample replaces `grep -qF` with an escaped regex wrapped in boundary character classes; an inline comment explains why grep -F alone is insufficient. Smoke test Step 9 (the `.bak` fixture) will now correctly produce `TOOL_FIX=no`.

File: docs/specs/developments/20260417154720_199-same-batch-tool-fix-ordering/2_199-same-batch-tool-fix-ordering_implementation-plan.md [hunk 5]
```diff
@@ -0,0 +1,254 @@
...
+  ".ai-dev-workflow.yaml"
+)
+PROTOCOLS_PREFIX="docs/ai/development-workflow/protocols/"
+
+classify_tool_fix() {
+  local dev_path="$1"
+  local doc_file matched_paths tool_fix
+
+  # Find the first spec or plan markdown document in the folder
+  doc_file="$(find "$dev_path" -maxdepth 1 -name '*.md' | sort | head -1)"
```
[the remark this answers is not in the document]
Comment (lhpaul) [reply docs/specs/developments/20260417154720_199-same-batch-tool-fix-ordering/2_199-same-batch-tool-fix-ordering_implementation-plan.md]: Addressed in commit `0f58261` — code sample now scans ALL .md files in the development folder (builds a bash array from `find ... | sort`), so tool-fix references appearing only in the implementation plan are detected.

File: docs/testing/workflow/199-same-batch-tool-fix-ordering.smoke-test.md [hunk 6]
```diff
@@ -0,0 +1,267 @@
...
+mkdir -p .tmp/smoke-199/20200101000008_ac13-no-doc
+# Do NOT create any .md file in this folder
+WORKFLOW_SKIP_FETCH=1 ./scripts/development-workflow/workflow-batch-plan.sh \
+  .tmp/smoke-199/20200101000008_ac13-no-doc 2>/dev/null || true
+```
+
+**Expected result**:
+- Output contains `TOOL_FIX=unknown`
+
+### Step 9: AC3 — Exact-path matching (no false positives from superstrings)
```
[the remark this answers is not in the document]
Comment (lhpaul) [reply docs/testing/workflow/199-same-batch-tool-fix-ordering.smoke-test.md]: Addressed in commit — Step 9 title now reads "AC1 — Exact-path matching (no false positives from superstrings)", correctly mapping to the AC the test validates.

File: docs/specs/developments/20260417154720_199-same-batch-tool-fix-ordering/2_199-same-batch-tool-fix-ordering_implementation-plan.md [hunk 7]
```diff
@@ -0,0 +1,281 @@
...
+
+  if [ "${#matched_paths[@]}" -gt 0 ]; then
+    tool_fix="yes"
+    local IFS=','
+    printf '%s\n' "yes" "${matched_paths[*]}"
+  else
+    printf '%s\n' "no"
+  fi
+}
+```
```
[the remark this answers is not in the document]
Comment (lhpaul) [reply docs/specs/developments/20260417154720_199-same-batch-tool-fix-ordering/2_199-same-batch-tool-fix-ordering_implementation-plan.md:260]: Acknowledged and deferred — the caller integration example is a nice-to-have for the code sample. The plan text at lines 55-67 already describes the call placement ("before the workflow-next-action.sh call") and the emission contract (TOOL_FIX always, TOOL_FIX_FILES only when yes). An implementer has enough information to wire in the call; the example polish can be added during implementation if needed. Closing to cap review churn at cycle 5.

File: docs/specs/developments/20260417154720_199-same-batch-tool-fix-ordering/2_199-same-batch-tool-fix-ordering_implementation-plan.md [hunk 8]
```diff
@@ -0,0 +1,281 @@
...
+)
+PROTOCOLS_PREFIX="docs/ai/development-workflow/protocols/"
+
+classify_tool_fix() {
+  local dev_path="$1"
+  local doc_file matched_paths tool_fix
+
+  # Find ALL spec and plan markdown documents in the folder (spec is 1_*_specs.md, plan is
+  # 2_*_implementation-plan.md). Scan every file, not just the first — a tool-fix reference
+  # may appear only in the implementation plan, and scanning only the spec would miss it.
```
[the remark this answers is not in the document]
Comment (lhpaul) [reply docs/specs/developments/20260417154720_199-same-batch-tool-fix-ordering/2_199-same-batch-tool-fix-ordering_implementation-plan.md:214]: Acknowledged — the comment does mention spec/plan naming specifically but the code scans all `*.md` files. This is intentional (matches plan line 37 "any \*.md file") and the paragraph's main point about scanning every file not just the first is correct. Rewording is pure polish — closing to cap review churn at cycle 5.

File: docs/testing/workflow/199-same-batch-tool-fix-ordering.smoke-test.md [hunk 9]
```diff
@@ -0,0 +1,267 @@
...
+cat > .tmp/smoke-199/20200101000009_ac3-superstring/1_ac3b_specs.md <<'EOF'
+# References scripts/development-workflow/pr-review-loop.sh.bak (a superstring, not the canonical path)
+EOF
+WORKFLOW_SKIP_FETCH=1 ./scripts/development-workflow/workflow-batch-plan.sh \
+  .tmp/smoke-199/20200101000009_ac3-superstring 2>/dev/null || true
+```
+
+**Expected result**:
+- Output contains `TOOL_FIX=no` (superstring must not trigger a match)
+
```
[the remark this answers is not in the document]
Comment (lhpaul) [reply docs/testing/workflow/199-same-batch-tool-fix-ordering.smoke-test.md:185]: Acknowledged — AC1 is also tested positively in Step 1; Step 9 is indeed a negative/boundary case for the exact-path matching requirement. The relabel from AC3 to AC1 is still more correct than leaving the AC3 mislabel (which was actively misleading). A further rename to "boundary validation" is pure polish — closing to cap review churn at cycle 5.

File: docs/specs/developments/20260417154720_199-same-batch-tool-fix-ordering/2_199-same-batch-tool-fix-ordering_implementation-plan.md [hunk 10]
```diff
@@ -0,0 +1,281 @@
...
+  block that emits `TARGET`, `DEVELOPMENT_PATH`, `SLUG`, and the TOOL_FIX line(s) before
+  continuing to the next path. This keeps spec AC 13 (`TOOL_FIX=unknown` for directories
+  without spec/plan documents) satisfiable. Emit rules for the TOOL_FIX lines (applies in
+  both the happy path and the next-action-failed path):
+  - `print_kv TOOL_FIX "$tool_fix"` always.
+  - `print_kv TOOL_FIX_FILES "$tool_fix_files"` only when `TOOL_FIX=yes` (omit the line
+    entirely for `no` and `unknown` to keep output compact).
+
+- [ ] The canonical tool file list to match against (exact string for each, checked with
+  `grep -qF` individually, plus the glob-equivalent grep for the protocols directory):
```
[the remark this answers is not in the document]
Comment (lhpaul) [reply docs/specs/developments/20260417154720_199-same-batch-tool-fix-ordering/2_199-same-batch-tool-fix-ordering_implementation-plan.md:79]: Acknowledged — plan step at line 78 still references `grep -qF` after lines 40-52 were updated to require delimiter-aware matching. This is an internal inconsistency from a prior fix cycle. The code sample (lines 205-245) is the authoritative implementation reference and uses the correct delimiter-aware pattern; an implementer following the code sample will do the right thing. Closing to cap review churn at cycle 5; worth a follow-up polish PR if the inconsistency misleads anyone during implementation.

File: docs/testing/workflow/199-same-batch-tool-fix-ordering.smoke-test.md [hunk 11]
```diff
@@ -0,0 +1,267 @@
...
+- [ ] AC7: no tool references → `TOOL_FIX=no` (explicit line emitted, not omitted)
+- [ ] AC8: tracker-derived conservative override described in Protocol 90 Step 3
+- [ ] AC9: serialize-first rule described in Protocol 90 Step 3
+- [ ] AC10: multiple tool-fix items serialized one at a time, documented in Protocol 90 Step 3
+- [ ] AC11: standard priority order for multiple tool-fix items documented
+- [ ] AC12: hazard explicitly named in Protocol 90 Step 3
+- [ ] AC13: missing spec/plan document → `TOOL_FIX=unknown`
+- [ ] AC14: `TOOL_FIX=unknown` treated as `yes` (serialize-first), documented in Protocol 90 Step 3
+- [ ] Superstring non-match: `pr-review-loop.sh.bak` does NOT trigger `TOOL_FIX=yes`
+- [ ] Step 5.1 in Protocol 90 is unchanged
```
[the remark this answers is not in the document]
Comment (lhpaul) [reply docs/testing/workflow/199-same-batch-tool-fix-ordering.smoke-test.md:239]: Acknowledged — assertions checklist is missing an explicit line for spec AC14 (human override). Step 5 verifies "Protocol 90 Step 3 contains all required content" which includes AC14, but a dedicated checkbox would be clearer. Deferring to cap review churn at cycle 5.

**Comment by lhpaul** [comment]
## Automated Reviewer Loop Summary

**Status**: All automated reviewers clean on HEAD `3de3486`; 5 deferred nits (all 🔵/🟡 Nitpick/Minor) closed via reply-only resolution with rationale.

| Platform | Result | Notes |
|---|---|---|
| Devin Review | ✅ Clean | SUCCESS commit status on `3de3486`. Earlier cycles: 2 🟡 findings addressed inline (classify_tool_fix placement, multiple tool-fix ACs), plus 2 🟡 findings on later cycles (superstring matching in plan+code, head -1) addressed. Final cycle raised 2 more 🟡 nits — closed with reply-only rationale to cap review churn at cycle 5. |
| CodeRabbit | ✅ Clean | SUCCESS commit status on `3de3486`. Earlier cycles: 1 🟠 Major (superstring matching) and 1 🟡 Minor addressed inline. Final cycle raised 3 🔵 Trivial nits — closed with reply-only rationale. |

**Review threads**: 0 unresolved (16/16 resolved — 6 Devin + 10 CodeRabbit, including 5 reply-only deferrals on trivial polish).

**CI**: All required checks passing (Markdown lint SUCCESS, E2E placeholder SKIPPED as expected for a plan branch).

**Notes on run path**: Item-orchestrator subagent terminated from an API socket error (`ConnectionRefused`) mid-run. Portfolio Orchestrator resumed inline across 5 review cycles (markdown-lint path fix → classify_tool_fix placement → 3-finding cycle 3 → AC1 label → 5 nit cycle 5). User invoked an explicit "accept-remaining-nits" decision at cycle 5 to cap churn. The 5 deferred findings are all 🔵 Trivial or 🟡 Minor polish (no blocking issues) — they can be addressed by small follow-up PRs or during implementation if the nits cause confusion. Captured in `project_batch12_retro_notes.md`.
